src/Control/Carrier/Class.hs

```diff
@@ -1,26 +1,49 @@
-{-# LANGUAGE FlexibleInstances, FunctionalDependencies #-}
+{-# LANGUAGE DeriveFunctor, EmptyCase, FlexibleInstances, FunctionalDependencies, TypeOperators, UndecidableInstances #-}
 module Control.Carrier.Class
 ( Carrier(..)
 ) where
 
-import Control.Effect.Choose (Choose(..))
+import {-# SOURCE #-} Control.Effect.Choose (Choose(..))
 import Control.Effect.Class
-import Control.Effect.Empty (Empty(..))
-import Control.Effect.Error (Error(..))
-import Control.Effect.NonDet (NonDet)
-import Control.Effect.Reader (Reader(..))
+import {-# SOURCE #-} Control.Effect.Empty (Empty(..))
+import {-# SOURCE #-} Control.Effect.Error (Error(..))
+import {-# SOURCE #-} Control.Effect.Lift (Lift(..))
+import {-# SOURCE #-} Control.Effect.NonDet (NonDet)
+import Control.Effect.Pure
+import {-# SOURCE #-} Control.Effect.Reader (Reader(..))
+import {-# SOURCE #-} Control.Effect.State (State(..))
 import Control.Effect.Sum ((:+:)(..))
-import Control.Effect.Writer (Writer(..))
-import Control.Monad ((<=<))
+import {-# SOURCE #-} Control.Effect.Writer (Writer(..))
+import Control.Monad ((<=<), join)
+import Data.Functor.Identity
+import qualified Control.Monad.Trans.Except as Except
+import qualified Control.Monad.Trans.Reader as Reader
+import qualified Control.Monad.Trans.RWS.CPS as RWS.CPS
+import qualified Control.Monad.Trans.RWS.Lazy as RWS.Lazy
+import qualified Control.Monad.Trans.RWS.Strict as RWS.Strict
+import qualified Control.Monad.Trans.State.Lazy as State.Lazy
+import qualified Control.Monad.Trans.State.Strict as State.Strict
+import qualified Control.Monad.Trans.Writer.CPS as Writer.CPS
+import qualified Control.Monad.Trans.Writer.Lazy as Writer.Lazy
+import qualified Control.Monad.Trans.Writer.Strict as Writer.Strict
 import Data.List.NonEmpty (NonEmpty)
 import qualified Data.Semigroup as S
+import Data.Tuple (swap)
 
 -- | The class of carriers (results) for algebras (effect handlers) over signatures (effects), whose actions are given by the 'eff' method.
 class (HFunctor sig, Monad m) => Carrier sig m | m -> sig where
   -- | Construct a value in the carrier for an effect signature (typically a sum of a handled effect and any remaining effects).
   eff :: sig m a -> m a
 
 
+-- base
+
+instance Carrier (Lift IO) IO where
+  eff = join . unLift
```

Comment:
This is pretty obvious in hindsight; I expect it’ll obviate the need for the vast majority of `runM`s.